File: app/src/main/java/com/labessence/biotech/finaldilution/component/MilimolarConcentration.java
package com.labessence.biotech.finaldilution.component;

/**
 * Project: FinalDilution
 * Created by dawid.chmielewski on 11/11/2017.
 */

class MilimolarConcentration extends Concentration {
    public MilimolarConcentration(double amount) {
        super(amount, ConcentrationType.MILIMOLAR);
    }

    @Override
    double calcDesiredMass(double volume, double molarMass) {
        return concentration * volume * molarMass / 1000 / 1000;
    }

    @Override
    double calcVolumeForDesiredMass(double mass, double molarMass) {
        return mass / molarMass / concentration * 1000 * 1000;
    }
}

File: app/src/main/java/com/labwhisper/biotech/finaldilution/component/concentration/MilimolarConcentration.kt
package com.labwhisper.biotech.finaldilution.component.concentration

import com.labwhisper.biotech.finaldilution.compound.NoMolarMassException

class MilimolarConcentration(amount: Double) : Concentration(amount, ConcentrationType.MILIMOLAR) {

    override val multiplicationFactor = 1e6

    override fun calcDesiredMass(volume: Double, molarMass: Double?): Double {
        if (molarMass == null) {
            throw NoMolarMassException()
        }
        return concentration * volume * molarMass / multiplicationFactor
    }

    override fun calcVolumeForDesiredMass(mass: Double, molarMass: Double?): Double {
        if (molarMass == null) {
            throw NoMolarMassException()
        }
        return mass / molarMass / concentration * multiplicationFactor
    }
}
